### PR TITLE
lime3ds-gui: Add Custom Layout interface

### PR DESCRIPTION
**Pull request by rtiangha**
This is another PR that was submitted for consideration (this time by [blakbin](https://github.com/blakbin/citra/commits/custom_layout/)) but never had a chance to make it into Canary or Nightly before the shutdown. The consensus seemed to be that it’d be merged in once the Configuration UI received a general overhaul, but that never had a chance to happen.

It gives an interface to the hidden Custom Layout functionality where you can explicitly specify the placement of the two screens. The functionality to do so was always there and could be specified via one of the text configuration files; this PR just gives the user a graphical way to do so.

As an example (see below), you could make one screen really big and the other really small, and even put one on top of the other and adjust the opacity of that second screen (unfortunately, customizable opacity is OpenGL only).

Some screenshots:

Configuration Interface
configuration

Example: Picture-in-Picture placement with second screen at 75% opacity
pip

Position values are in pixels starting from (0,0) (i.e. the top left of the display), which are the units that the original text-based functionality uses. Opacity is in percent.

Original Description:

> This add Custom Layout selector in Both Menu and Enhancement Setting. This will allow to use or change Custom Layout mode or by using toggle no matter what config 'custom_layout\default' value true or false. Still No Overlay Editor, just manual edit box, and this Not saved or carried at per game settings.

[Original PR](https://web.archive.org/web/20240107122308/https://github.com/citra-emu/citra/pull/6532)

### CODE REVIEW

**Comment by Gamer64ytb** [comment]
I think that we should add a layout editor before just a toggle as well, but I don't think it's a easy task so idk

**Comment by OpenSauce04** [comment]
> I think that we should add a layout editor before just a toggle as well

Not sure what you mean, this isn't just a toggle. The menu itself is the editor.

If you are referring to a graphical editor where you can drag the windows around and stuff, that would likely require a large amount of work, and isn't an immediate concern when compared to the immediate benefits this implementation gives in its current state.

**Comment by OpenSauce04** [comment]
Given that this new feature doesn't really qualify as an 'enhancement' and takes up quite a large space in the menu, I think it would be beneficial to move this to a new tab between Enhancements and Advanced named "Layout". What are your thoughts on this?

**Comment by rtiangha** [comment]
Good thing I can see the future, lol. I'll push things out once I've confirmed they compile on this end. Shouldn't be too long since I've already done the work.

**Comment by OpenSauce04** [comment]
I will hopefully get around to reviewing all of your PRs on Saturday and Sunday. Thank you for all of the contributions.

**Comment by rtiangha** [comment]
And done. New screenshots:

Enhancements Tab
enhancements

Layout Tab
Layout

I took the opportunity to reorganize the positioning fields since this way made more sense to me. But if people want the old order (or a different one), let me know.

File: src/lime_qt/configuration/configure_enhancements.ui [hunk 1]
```diff
@@ -308,168 +308,13 @@
...
-           <item>
-            <property name="text">
-             <string>Single Screen</string>
-            </property>
-           </item>
-           <item>
-            <property name="text">
-             <string>Large Screen</string>
-            </property>
-           </item>
```
Comment (OpenSauce04) [inline src/lime_qt/configuration/configure_enhancements.ui:355]: The 'Hybrid Screen' layout option is missing from the Layout configuration menu

**Review by OpenSauce04 — CHANGES_REQUESTED** [review]

**Comment by OpenSauce04** [comment]
The 'Hybrid Screen' view option appears to be missing from the Layout configuration tab 
image
Selecting Hybrid Screen in the View menu causes Custom Layout to be selected in this menu, and selecting Custom Layout in the View menu causes no option to be selected in this menu:
image

**Comment by rtiangha** [comment]
OK, I think that should do it, but please verify. I also took the opportunity to reorder the View menu to match the Layout drop down order to make things more consistent and to satisfy my OCD.

**Comment by OpenSauce04** [comment]
This PR functions fine, however I would like to change how this functionality is implemented from a UX perspective.
While the currently implemented Top, Bottom, Left, Right setup works, it is not particularly intuitive. I would like to swap this out for a conventional X, Y, Width, Height setup, as this is much easier to understand and manipulate than setting absolute positions for each of the screen's edges.

I am willing to implement this, however please disable required commit signing for your fork so I am able to push changes. I am not willing to set up commit signing as I work across several different machines, and it is not worth my time.

**Comment by rtiangha** [comment]
Have at 'er.

**Comment by OpenSauce04** [comment]
image
What is this value supposed to do exactly? I can't seem to get any discernible effect out of changing it

**Comment by OpenSauce04** [comment]
Oh, it's a value exclusive to the Large Screen view option, and changes the size each screen takes up. This should be greyed out when other layouts are selected to avoid confusion. I will implement this change along with my previously mentioned amendments

**Comment by rtiangha** [comment]
It was there before the patch. That said, I think for at least one of the layout modes (can't remember which one(s)), it affects how large the main screen is relative to the secondary screen. I think it's supposed to be a convenience thing and was the only way for the user to resize screens before this patch. Would have no effect on the Custom Layout profile since it uses different variables.

Edit: Too slow, lol. But yes; would be easy to grey out if one of the other modes were selected. There are probably a bunch of things in the Configure UI that should be that way, lol.

**Comment by OpenSauce04** [comment]
I appear to still be unable to push my changes to your fork, please check the access permissions
image

**Comment by rtiangha** [comment]
Try it now.

**Comment by OpenSauce04** [comment]
New UI layout:
image

**Comment by rtiangha** [comment]
Yeah, that is better.

**Comment by OpenSauce04** [comment]
As I have made notable changes to this PR, I am going to request that you also review the changes to get a second pair of eyes on my changes
As you created the PR, you cannot make a formal review, so please just state your opinion as a commit

**Review by OpenSauce04 — APPROVED** [review]

**Comment by rtiangha** [comment]
LGTM, although I am a bit curious about why the need to cast to (u32). The values are defined as u16 to begin with; isn't 32-bits for something that's probably max 4 digits long a bit overkill? And even if there is a monitor out there that's larger than 65,535 pixels, wouldn't it be better to define them as u32 to begin with? Actually, I'm not sure what casting from u16 to u32 does, except double the memory usage for an integer that doesn't change. What's the purpose?

**Comment by OpenSauce04** [comment]
The cast is simply to stop the compiler from whingeing when adding the two numbers together.
This is what happens when the cast is not present:
```
/var/host/home/daniel/repos/Lime3DS/src/core/frontend/framebuffer_layout.cpp:382:80: error: narrowing conversion of '(((int)Settings::values.Settings::Values::custom_top_x.Settings::Setting<short unsigned int>::GetValue()) + ((int)Settings::values.Settings::Values::custom_top_width.Settings::Setting<short unsigned int>::GetValue()))' from 'int' to 'unsigned int' [-Werror=narrowing]
```
u32 was chosen instead of u16 to avoid a case where both a very large X/Y and Width/Height are added together and the number overflows. This could probably be fixed by simply reducing the maximum value for these configs, but it's really not that big of a deal either way.
~~This isn't the 90s, we can spare a few bytes of wasted RAM 😛~~ Actually, these numbers are cast to 32 bits anyway because of the type of the Rectangle struct
image

**Comment by rtiangha** [comment]
Ah, got it. Ok, definitely LGTM.

**Pull request merged**